### ISSUE TITLE
Erro 409 quando cnpj é válido.

### ISSUE THREAD

**Issue by pedrro** [ThoughtWorksInc/registrolivre#122]
O erro 409 é mostrado no console quando um cnpj é validado pela diretiva.

Comportamento esperado:
Se um cnpj for válido, deve retornar  error 200
Se um cnpj for duplicado, deve retornar error 409

**Comment by pedrro** [comment]
relacionado a estória #100

**Comment by waldemarnt** [comment]
O front end faz um GET para buscar-por-cnpj, levando em consideração a verbosidade ele esta pegando uma company por CNPJ, no meu ponto de vista o comportamento esperado é 200 para sucesso e 404 caso não seja encontrada nenhuma company com este CNPJ.
O que tu acha?

**Comment by pedrro** [comment]
Mas ai, se não houver uma compania com o CNPJ, o campo vai validar e mostrar 404 no console, acha certo? 
 Deixa eu mostrar alguns cenários para explicar melhor quando os status code devem ser enviados:

**O back-end tem que enviar o erro 200 para a diretiva quando:**
Usuário adicionou no campo de CNPJ o seguinte valor: 000.000.000/0001-91
O back-end valida e retorna para a diretiva o status code 200 pois o CNPJ não existe na base de dados.
A diretiva adiciona no campo a classe de has-sucess.

**O back-end tem que enviar o erro 409 para a diretiva quando:**
Usuário adicionou no campo de CNPJ o seguinte valor: 000.000.000/0001-91
O back-end valida e retorna para a diretiva o status code 409 pois o CNPJ já existe na base de dados.
A diretiva então exibe a mensagem de erro no campo: Já existe uma empresa com esse CNPJ cadastrada.  adiciona a classe has-error.

**404 tem que ser quando o CPNJ está em um formato invalido, exemplo:**
Usuário adicionou no campo de CNPJ o seguinte valor: 111.111.111/1111-11
O back-end retorna erro 400 pois o CNPJ é invalido
A diretiva exibe a mensagem de validação "Número Invalido" no campo e adiciona a classe has-error.

**Comment by waldemarnt** [comment]
Poise, são duas coisas diferentes, o nosso metodo de getCompanyByCnpj como o nome diz ele busca uma compania por cnpj, se achar ela, o correto é 200 se não achar o certo é 404 (esse é o comportamento esperado pelo nosso método).

O front precisa saber de 3 coisas, se o CNPJ é válido, se for invalido o código correto é 422 (entidade não pode ser processada), se for válido é 200.
Se nada for encontrado o código é 404 
Se for é 200

Como a nossa comunicação é como uma API não tem problema o front end receber um 404 e tratar esse erro, na verdade é o caso esperado né?

Ou podemos fazer outro método que retorne um booleano apenas para válidar o CNPJ.

**Comment by pedrro** [comment]
**Como a nossa comunicação é como uma API não tem problema o front end receber um 404 e tratar esse erro, na verdade é o caso esperado né?**

É o esperado porque foi modelado desse jeito, que eu considero errado, porque não me parece correto a gente retornar um status code 404 para uma coisa que é valida/que pode ser usada. Os status code de série 4xx são usados justamente quando alguma coisa está errada, seja um request, um método que não existe, um dado não aceitável e etc... Essa é a minha justificativa de porque acho que essa modelagem está errada. 
ps: Talvez o status code 200 nem seja tão indicado para esse cenário, acho que o 202 parece ser mais válido.

O que eu acho que seria valido ser feito para esse bug, é refatorar como a diretiva trata a resposta do back-end para utilizar dos status code certos.

**Comment by waldemarnt** [comment]
O client nao pode por regra sobre a API, a nossa api, na chama de getCompanyByCnpj deve retorna 200 de OK(sucesso) ou 404 de not found.
O cliente quer ver se uma company existe na API e recebe um 404 que não existe não é errado é a resposta correta da api.
O status 202 é um sucesso, ele quer dizer que tu mandou uma request pra API, a request foi aceita e validada e vai ser processada em background, normalmente usada pra coisas assincronas.
Acho que a gente ta tendo um erro de conceito aqui, pois mesmo refatorando o client(front end) vai precisar verificar se o cnpj existe ou não, e vamos cair neste dilema novamente xD

**Comment by brunapereira** [comment]
Temos no back um método chamado "getCompanyByCnpj".
Esse método retornou uma Company?
R1: Sim ---> Retorna 200 Success.                           R2: Não -->
Retorna 404 Not Found.

Alterar o método de getCompanyByCnpj para retornar Success quando NÃO pegar
a empresa (que é a única coisa que ela tem que fazer) seria bem
contraditório.

Dito isso temos 2 opções:

1: Criar exatamente o mesmo método COM OUTRO NOME mas que retorna um valor
booleano. Ex:
isCompanyInDatabase()
R1: Sim ---> Retorna True.                                      R2: Não -->
Retorna False.

2: Deixar do jeito que tá.

**Comment by josigama** [comment]
Concordo com Waldemar e a Bruna, pois fica muito mais intuitivo quando a API acha uma empresa que a resposta seja de sucesso (200) e quando não acha seja not found (404). 

Em vez de colocarmos nossas opiniões em comentários, podemos sentar amanhã e compartilhar com o time, chegando a um consenso.

**Comment by dieselpardal** [comment]
Revisar.
